src/IO/H5/Cce.cpp

```diff
@@ -0,0 +1,262 @@
+      }
+    }
+    {
+      const htri_t header_exists = H5Aexists(cce_group_.id(), "header.hdr");
+      if (header_exists != 0) {
+        const Header header(true, detail::OpenGroup{}, cce_group_.id(),
+                            "header");
+        header_ = header.get_header();
+      }
+    }
+
+    for (const std::string& bondi_var : bondi_variables_) {
+      DataSet& dataset = bondi_datasets_.at(bondi_var);
+      dataset.id =
+          H5Dopen2(cce_group_.id(), bondi_var.c_str(), h5::h5p_default());
+      CHECK_H5(dataset.id, "Failed to open dataset");
+
+      hid_t space_id = H5Dget_space(dataset.id);
+      std::array<hsize_t, 2> max_dims{};
+      if (2 != H5Sget_simple_extent_dims(space_id, dataset.size.data(),
+                                         max_dims.data())) {
+        ERROR("Invalid number of dimensions in cce file " << bondi_var
+                                                          << " on disk.");
+      }
+      CHECK_H5(H5Sclose(space_id), "Failed to close dataspace");
+
+      if (legend_ != read_rank1_attribute<std::string>(dataset.id, "Legend"s)) {
+        ERROR("l_max from cce file " << bondi_var
+                                     << " does not match l_max in constructor");
+      }
+    }
+  } else {  // file does not exist
+    {
+      Version open_version(false, detail::OpenGroup{}, cce_group_.id(),
+                           "version", version_);
+    }
+    {
+      Header header(false, detail::OpenGroup{}, cce_group_.id(), "header");
+      header_ = header.get_header();
+    }
+    // So this file is compatible with the sxs/scri python packages offered by
+    // the SXS collaboration
+    write_to_attribute(cce_group_.id(), sxs_format_str_, sxs_version_str_);
+
+    for (const std::string& bondi_var : bondi_variables_) {
+      DataSet& dataset = bondi_datasets_.at(bondi_var);
+      dataset.id = h5::detail::create_extensible_dataset(
+          cce_group_.id(), bondi_var, dataset.size,
+          std::array<hsize_t, 2>{{4, legend_.size()}},
+          {{h5s_unlimited(), legend_.size()}});
+      CHECK_H5(dataset.id, "Failed to create dataset");
+
+      write_to_attribute(dataset.id, "Legend"s, legend_);
+      // So this dataset is compatible with the sxs/scri python packages offered
+      // by the SXS collaboration
+      write_to_attribute(dataset.id, sxs_format_str_, sxs_version_str_);
+    }
+  }
+}
+
+Cce::~Cce() {
+#ifdef __clang__
+  for (const auto& name_and_dataset : bondi_datasets_) {
+    const auto& name = name_and_dataset.first;
+    const auto& dataset = name_and_dataset.second;
+    CHECK_H5(H5Dclose(dataset.id), "Failed to close dataset " << name);
+  }
+#else
+// The pragma here is used to suppress the warning that the compile time branch
+// of `ERROR` will always call `terminate()` because it throws an error.
+// Throwing that error is a code path that will never actually be entered at
+// runtime, so we suppress the warning here.
```

Comment:
Why can't it happen?

Reply:
To be honest, I just copied this from the `Dat.cpp` file, so not really sure. @macedo22 added this when she made the ERROR macros usable at compile time.

Reply:
The warning being suppressed is for the branch taken in `ERROR` when we have a constant evaluated context (the code is being evaluated at compile time). This branch of `ERROR`, by construction, will never run at runtime because in this branch, we will get the error thrown at compile time.

Reply:
See the branch [here](https://github.com/sxs-collaboration/spectre/blob/a9ee6aa7c65ba703dbdc6dfb79436bb5378d4465/src/Utilities/ErrorHandling/Error.hpp#L68)

Reply:
At least, that is my understanding from when we made this change